Apply tokens and Execute Token (#39)
plantanapp/documentation: docs/Actions/apply-tokens.md

---
id: apply-tokens
title: Apply Tokens
sidebar_label: Apply Tokens
---

> Audience: [`Citizen Developer`](/docs/audience#citizen-developers)<br/>

> > Skill Prerequisites: `Variables`

Author: OK done

> This actions executes multiple expressions and applies tokens on those expressions and stores it in another (or the same) token(s). Note that this presents a security risk when the expression has content from untrusted users. 

Author: Added security section and explanation there

## `Typical Use Cases`

> - Execute a Variable expression that returns an token that needs to be reevaluated

Author: Execute a token expression (a mix of tokens and/or free text) that returns an other token expression that needs to go thru the same process

- Apply tokens inside a text from various sources (eg: other actions, form fields, API input parameters)

## `Don't use it to`

> - Concatenate 2 tokens - Use [Inject Data](/docs/Actions/inject-data.md)

Author: There is no reason. I was trying to encourage people to try more actions. Actually Inject Data, Execute Token and Apply Tokens can be one action.

> - Overwrite a token - Use [Inject Data](/docs/Actions/inject-data.md)

Reviewer: Got it. But why shouldn't I use this action instead?

Author: There is no reason. I was trying to encourage people to try more actions. Actually Inject Data, Execute Token and Apply Tokens can be one action.

## `Related Actions`

| Action Name                                                     | Description                                                                        |
| --------------------------------------------------------------- | ---------------------------------------------------------------------------------- |
| [Update form with AJAX](/docs/Actions/update-form-with-ajax.md) | Form module only, it updates the fields based on the values from the tokens.       |

> | [Display Message](/docs/Actions/display-message.md)             | Form Module only, shows a message after the executions of set of actions/workflow. |

Author: Added Run SQL good point. Actually, this works with all, I've tried to just show some more specific ones from real cases.

## `Input Parameter Reference`

| Parameter                 | Description                                                                                                                 | Supports Variables |
| ------------------------- | --------------------------------------------------------------------------------------------------------------------------- | ------------------ |
| Token -> Token Expression | The tokens and/or text (expression) you want to be evaluated or reevaluated                                                 | Yes                |

> | Number of Recursions      | Specify how many times to apply tokenization. This is useful when tokens contain other tokens that also need to be replaced | Yes                |

Reviewer: What is the default?

Author: Added default

## `Output Parameters Reference`

| Parameter         | Description                                                                 |
| ----------------- | --------------------------------------------------------------------------- |
| Token -> Store as | The variable name(s) in which the response is to be stored for further use. |

## `What is recursion and when to use it`

Recursion refers of how many times we should try to check for tokens in the expression and try to replace them. 

If recursion is set to 1 then this action will work as [Execute Token](/docs/Actions/execute-token.md) action. Will assign the Expression to a token.

> There are situation ([like example number 2](#2-apply-tokens-inside-a-template-recursion-1)) where you save a template (let's say email template) and you save it with tokens instead of the First Name and Last Name so you can save only one copy and when you need to deliver that email you only have to replace the Variables. 

Reviewer: Huh!?
Let's rephrase this, it's hard to understand and avoid using the "save" word 3 times in a single sentence.
Also, in English, prefer shorter sentences.

Author: Made it in shorter sentences.

> Eg. Email Template: Hi, [User:FirstName] [User:LastName]. Please contact me!

Reviewer: Let's place token names inside code tags so they look nicer. E.g. using the back ticks `[User:FirstName]`

Reviewer: Or, place the entire example in a code block.

Author: added ticks

Let's say we have the above template stored in \[EmailTemplate\]

> When will bring the template value will have to set it's value in a token that will contain tokens itself so we will need to set the recursion to 2 to replace the token to the template and the template to a plain token. See below example for Recursion set to 2:

Reviewer: Split in multiple sentences so it's easier to follow. Also, will = we'll and it's  = its?

Author: Simplified it

> \[EmailTemplate\] -> Hi, \[User:FirstName\] \[User:LastName\]. Please contact me! -> Hi, John Doe. Please contact me!

Reviewer: Same here, let's add code blocks for examples.

Author: added ticks

## `Examples`

> ### `1. Calculations with Variable DoMath (Recursion =1)`

Reviewer: Let's call it Token instead of Variable

Author: Replaced Variables to tokens in all md

> By default the expression will only concatenate the variables and/or text from the token expression so in case we need to do some math Variables provides a token called DoMath that can solve this issue.

Reviewer: ```suggestion
Applying tokens is just replacing the placeholders inside the token expression. If we needed to do some arithmetic operations, we can use the built-in [DoMath](/docs/tokens/do-math) token.
```

```json

{
    "Title": "Apply Tokens",
    "ActionType": "ApplyTokens",
    "Parameters": {
        "Tokens": [
            {
                "value": "[DoMath:Eval(expression=[User:UserId]*2)]",
                "name": "DoubleOfUserId"
            },
            {
                "value": "[DoMath:Eval(expression=(5*2)+2)]",
                "name": "DoubleOfFivePlusTwo"
            }
        ],
        "RecursionCount": "1"
    }
}
​
```

### `2. Apply tokens inside a template (Recursion >1)`

> This example will show you how to replace tokens inside another token.

Reviewer: ```suggestion
This example demonstrates replacing tokens inside the content returned by evaluating a different token named `[EmailTemplate]`.

```

                        {
                            "value": "Hi, [User:FirstName] [User:LastName]. Please contact me!",
                            "name": "EmailTemplate"
                        }
                    ],
                    "RecursionCount": "1"
                },
                "$_uid": "action15910976849431312",
                "$_isOpen": true,
                "$_isLoaded": true,
                "$_isFocus": true,
                "Description": "Assign Template to Token"
            },
            {
                "Title": "Apply Tokens",
                "ActionType": "ApplyTokens",
                "Parameters": {
                    "Tokens": [
                        {
                            "value": "[EmailTemplate]",
                            "name": "FinalTemplate"
                        }
                    ],
                    "RecursionCount": "2"
                },
                "$_uid": "action1591097684943656",
                "$_isOpen": true,
                "$_isLoaded": true,
                "$_isFocus": true,
                "Id": -1,
                "Description": "Apply tokens twice"
            }
        ]
    }
}

```

> ## `Frequently asked questions`

Reviewer: I like the questions that were addressed. However, seems like many of them apply to any action, not just this one. So we can pull them on separate pages.
Also, for the rest, I would try to integrate them into the section they belong to. I think the FAQs are very useful, but not on a documentation page where all should have already been covered. I see FAQ more of a summary of an extensive documentation. We can discuss having a separate FAQ section in the documentation that will live at learn.plantanapp.com/faq.

Author: The reality is that all the questions are generic. I will remove them and move them to the FAQ page and make those questions generic to apply to all the Actions

Author: I made a section for the "questions" that are action-specific related

> **What is the best Number of Recursion**

Reviewer: ```suggestion
**What is the average number of recursions?**

```

> Usually 2 or 3 depending of the situation should be enough but it really depends on what you need to achieve. Setting the number of recursion bigger that is need it will not have any impact on the result but will impact the performance. If you are not sure how recursion works increment the value by one until you get the desire output and there are no tokens left. That will be the desire value for the execution.

Reviewer: ```suggestion
It really depends on what you need to achieve. 2 or 3 recursions is the maximum that we've seen. Setting the number of recursion higher than needed will not have any impact on the result, but will impact the performance. If you are not sure how recursion works, increment the value by one until you get the desired output and there are no tokens left. That will be the ideal number for your use case.

```

**How long can I use the output token?**

> It can be used after the action is executed in the execution flow and/or Workflow and it be available until the executions ends or it is stopped. 

Reviewer: Avoid using execute 3 times in a sentence. Split in multiple sentences or find equivalent expressions.

**Can I use multiple actions in a row?**

> Yes. There is no limit or restriction regarding the usage of this action. Be aware that a bigger than need it Number of recursion may affect the performance.

Reviewer: Why would I use more actions in a row rather than putting everything in a single action?

Author: Maybe for some reason, you want to concatenate this output token to another and you can't do it in the same action or want to apply this output token as a param to another token.

**Where can I use the output token?**

> The output token can be used starting from the next action as any other token in conditions, as parameters to other actions etc. Store as values (output tokens) will not be available in the execution of the same action even if they are above in the token/expression list.

Reviewer: ```suggestion
The output token can be used starting with the next action. It's used just like any other token - in conditions, as parameters to other actions and so on. The output tokens will not be available in the same action where they are defined, even if they are above in the token/expression list.

```

**What tokens can I use?**

> Module related tokens (form fields, API input parameters etc.), Workflow input parameters, Variables tokens and other output tokens from the same execution/workflow.

Reviewer: ```suggestion
Module related tokens (form fields, API input parameters etc.), workflow input parameters, global tokens and tokens previously generated by other actions in the same process.
```
